### DISCUSSION TITLE
How can we allow file execution in home?

### DISCUSSION THREAD

**Discussion by blastrock** [netblue30/firejail#5228] [Q&A] [answered]
Hello,

It looks like when firejail hides the home directory, the mounted folder is somehow noexec (mount doesn't show it as such though...). I would like to be able to execute binaries from my private home.

```
# whitelist at list one file so that $HOME is hidden, I took `todo` here
$ firejail --noprofile --shell=none --whitelist=~/todo /bin/zsh
# now, inside the jail
$ cp /bin/ls .
$ ./ls
zsh: permission denied: ./ls
```

I tried adding `--ignore="noexec ${HOME}"`, it didn't help. Can you help me with that?

**Comment by ghost** [comment]
> $ firejail --noprofile --shell=none --whitelist=~/todo /bin/zsh

When you start a sandbox like that, `pwd` will be ${HOME}. Without doing something extra, `ls` is NOT copied to the whitelisted dir but to ${HOME}. You can either `cd` into ~/todo before running the cp command, or explicitly cp /bin/ls to ~/todo. Yet another option is to use Firejail's `--private-cwd=~/todo`, which will put you in the whitelisted dir once you enter the jail, avoiding extra steps with your example code.

Side note: not that there's anything wrong with using `--noprofile` here, but it might obscure things. If you drop that, Firejail will use its default.profile. That at least offers some minimal protections, while the former does not.

**Reply by blastrock** [reply]
Sorry, I have been unclear. I do not care about todo, I just want a private home. When you whitelist any file in $HOME, your home becomes private. There's probably a better option for that, but I used it only for the sake of the example.

Then, in that private home (or any subfolder, whitelisted or not), I want to be able to run executable files.

**Reply by rusty-snake** [reply]
The `tmpfs` is mounted `nosuid,nodev,noexec` as I can see it.

**Reply by blastrock** [reply]
Oh I see, I have

```
/dev/nvme0n1p3 on /home/blastrock type ext4 (rw,relatime,discard)
tmpfs on /home type tmpfs (rw,nosuid,nodev,noexec,mode=755,inode64)
```

I only read the first line before, but I think it comes from my system, and not from the jail. The real mount point comes from the second line as you say.

Anyway, can we remove that noexec?

**Reply by rusty-snake** [reply]
W/o patching? No.

**Reply by ghost** [reply]
> [...] I just want a private home. When you whitelist any file in $HOME, your home becomes private [...]

That description threw me off. There's a difference between `whitelisting` and Firejail's `--private`/`private-home=foo` options. I think @rusty-snake's comments relate to the latter.
One could create a sandbox with `--whitelist=${HOME}`, which would allow executing files under ${HOME}. But I wouldn't call it `private` in Firejail's sense of the term. Anyway, this is just me nitpicking.

**Reply by blastrock** [reply]
Yeah sorry for that, the issue is the same with `firejail --noprofile --shell=none --private /bin/zsh`, the home is mounted as noexec.

Shall I open a feature request for that then?

**Reply by ghost** [reply]
> Shall I open a feature request for that then?

When you do, it might be informational to refer to this discussion and make it clear you would like a `--private-exec` and/or `--private-home-exec=foo` option. The naming is just a suggestion, but I think that's what you actually want if I'm understanding all this correctly. Personally I would need more time to think about use cases for such a setup. You might want to specify yours more fully in a feature request.

**Reply by blastrock** [reply, marked as answer]
Done!

* https://github.com/netblue30/firejail/issues/5229
